diff --git a/apple/Package.swift b/apple/Package.swift
index 3eae728dd9962df808d9676f5f6171b53ea1a7fe..f6b9b4075761cbb31eda885ae6f5544754a8a958 100644
--- a/apple/Package.swift
+++ b/apple/Package.swift
@@ -134,7 +134,7 @@ let package = Package(
       dependencies: testFrameworks.dependencies,
     ),
   ] + testFrameworks.binaryTargets,
-  swiftLanguageModes: [.v6],
+  swiftLanguageModes: [.v5],
   cxxLanguageStandard: .cxx20
 )
 
diff --git a/apple/Sources/ExpoModulesJSI/Contexts/HostFunctionContext.swift b/apple/Sources/ExpoModulesJSI/Contexts/HostFunctionContext.swift
index bc89b9cf91825b4bd98f59bcaaf90bd4d4284257..7781dd2ce3db8f2c403ac8d2875a80ba9f39550c 100644
--- a/apple/Sources/ExpoModulesJSI/Contexts/HostFunctionContext.swift
+++ b/apple/Sources/ExpoModulesJSI/Contexts/HostFunctionContext.swift
@@ -2,7 +2,7 @@
  Context that captures Swift values to pass them to JSI host function as an unmanaged pointer for interoperability with C++.
  */
 internal final class HostFunctionContext: Sendable {
-  weak let runtime: JavaScriptRuntime?
+  nonisolated(unsafe) weak var runtime: JavaScriptRuntime?
   let name: String?
   let call: JavaScriptRuntime.SyncFunctionClosure
 
diff --git a/apple/Sources/ExpoModulesJSI/Contexts/HostObjectContext.swift b/apple/Sources/ExpoModulesJSI/Contexts/HostObjectContext.swift
index f61fcdae2de5e7f2b1272bf8d6283e1491e1e5f2..47f19f2fffedd9935c06d6fec89e4c7759174710 100644
--- a/apple/Sources/ExpoModulesJSI/Contexts/HostObjectContext.swift
+++ b/apple/Sources/ExpoModulesJSI/Contexts/HostObjectContext.swift
@@ -7,7 +7,7 @@ internal final class HostObjectContext: Sendable {
   typealias PropertyNamesGetter = @JavaScriptActor () -> [String]
   typealias Deallocator = @JavaScriptActor () -> Void
 
-  weak let runtime: JavaScriptRuntime?
+  nonisolated(unsafe) weak var runtime: JavaScriptRuntime?
   let get: Getter
   let set: Setter?
   let getPropertyNames: PropertyNamesGetter
diff --git a/apple/Sources/ExpoModulesJSI/Runtime/JavaScriptActor.swift b/apple/Sources/ExpoModulesJSI/Runtime/JavaScriptActor.swift
index 699c59a03fe883f280f484aa35ed6312317edfb3..60f958711b57a0fd0d07a652728d4680bab92183 100644
--- a/apple/Sources/ExpoModulesJSI/Runtime/JavaScriptActor.swift
+++ b/apple/Sources/ExpoModulesJSI/Runtime/JavaScriptActor.swift
@@ -93,7 +93,7 @@ internal class JavaScriptExecutor: SerialExecutor, @unchecked Sendable {
  An actor that is dedicated for the specific runtime.
  */
 internal actor JavaScriptRuntimeActor {
-  private weak let runtime: JavaScriptRuntime?
+  nonisolated(unsafe) private weak var runtime: JavaScriptRuntime?
   nonisolated private let executor: JavaScriptExecutor
 
   init(runtime: JavaScriptRuntime) {
@@ -114,7 +114,7 @@ internal actor JavaScriptRuntimeActor {
  Serial executor dedicated for the specific runtime.
  */
 internal final class JavaScriptRuntimeExecutor: JavaScriptExecutor, @unchecked Sendable {
-  private weak let runtime: JavaScriptRuntime?
+  nonisolated(unsafe) private weak var runtime: JavaScriptRuntime?
 
   init(runtime: JavaScriptRuntime) {
     self.runtime = runtime
diff --git a/apple/Sources/ExpoModulesJSI/Runtime/JavaScriptPropNameID.swift b/apple/Sources/ExpoModulesJSI/Runtime/JavaScriptPropNameID.swift
index b497d4fb5754394395680db77c2216310345caa0..4390c5b2e765db2f403dd55c47c8f843a539cab4 100644
--- a/apple/Sources/ExpoModulesJSI/Runtime/JavaScriptPropNameID.swift
+++ b/apple/Sources/ExpoModulesJSI/Runtime/JavaScriptPropNameID.swift
@@ -4,7 +4,7 @@ internal import jsi
  Represents something that can be a JS property key.
  */
 public final class JavaScriptPropNameID: JavaScriptType {
-  private weak let runtime: JavaScriptRuntime?
+  nonisolated(unsafe) private weak var runtime: JavaScriptRuntime?
   internal let pointee: facebook.jsi.PropNameID
 
   /**
diff --git a/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptArray.swift b/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptArray.swift
index 12acf78bcf80355a699bfe7e286fea5b0e9e1546..fa39171b4737854f952eed540fb4fd616dd64d39 100644
--- a/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptArray.swift
+++ b/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptArray.swift
@@ -7,7 +7,7 @@ internal import ExpoModulesJSI_Cxx
  to the underlying JavaScript array and provides Swift-friendly APIs for common array operations.
  */
 public struct JavaScriptArray: JavaScriptType, ~Copyable {
-  internal weak let runtime: JavaScriptRuntime?
+  nonisolated(unsafe) internal weak var runtime: JavaScriptRuntime?
   internal let pointee: facebook.jsi.Array
 
   /**
diff --git a/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptArrayBuffer.swift b/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptArrayBuffer.swift
index 72895224dde9c2a478a1b5c213176c5be07735a1..4152c1a6a412b2255e56a8a682f95469c7d4475f 100644
--- a/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptArrayBuffer.swift
+++ b/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptArrayBuffer.swift
@@ -8,7 +8,7 @@ internal import ExpoModulesJSI_Cxx
  underlying raw data pointer and size of the buffer.
  */
 public struct JavaScriptArrayBuffer: ~Copyable {
-  internal weak let runtime: JavaScriptRuntime?
+  nonisolated(unsafe) internal weak var runtime: JavaScriptRuntime?
   internal let pointee: facebook.jsi.ArrayBuffer
 
   internal init(_ runtime: JavaScriptRuntime, _ arrayBuffer: consuming facebook.jsi.ArrayBuffer) {
diff --git a/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptBigInt.swift b/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptBigInt.swift
index c9e68e590d84fb3b4190a1396531e6456ad7fb51..5afaa4d4f1f05109ce11b7af607f77bb318f72a6 100644
--- a/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptBigInt.swift
+++ b/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptBigInt.swift
@@ -39,7 +39,7 @@ internal import jsi
  You must explicitly convert between them.
  */
 public struct JavaScriptBigInt: JavaScriptType, Sendable, ~Copyable {
-  internal weak let runtime: JavaScriptRuntime?
+  nonisolated(unsafe) internal weak var runtime: JavaScriptRuntime?
   internal var pointee: facebook.jsi.BigInt
 
   /**
diff --git a/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptError.swift b/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptError.swift
index 79a2fbbc7cfe8a4797a24d9f9838ac3f2c3cfcf3..fed8de5e774f8e10baba05eef010772b2e0e6b19 100644
--- a/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptError.swift
+++ b/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptError.swift
@@ -2,7 +2,7 @@ internal import jsi
 internal import ExpoModulesJSI_Cxx
 
 public struct JavaScriptError: JavaScriptType, ~Copyable {
-  private weak let runtime: JavaScriptRuntime?
+  nonisolated(unsafe) private weak var runtime: JavaScriptRuntime?
   nonisolated(unsafe) private let pointee: facebook.jsi.JSError
 
   internal init(_ runtime: JavaScriptRuntime, _ error: facebook.jsi.JSError) {
diff --git a/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptFunction.swift b/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptFunction.swift
index 53c244579fac7dc6c8d265d46cd1b82f020ef4dc..79313badc2357bbf1e8dd68106413e2351caaa7a 100644
--- a/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptFunction.swift
+++ b/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptFunction.swift
@@ -4,7 +4,7 @@ internal import jsi
 internal import ExpoModulesJSI_Cxx
 
 public struct JavaScriptFunction: JavaScriptType, ~Copyable {
-  internal weak let runtime: JavaScriptRuntime?
+  nonisolated(unsafe) internal weak var runtime: JavaScriptRuntime?
   internal let pointee: facebook.jsi.Function
 
   internal init(_ runtime: JavaScriptRuntime, _ pointee: consuming facebook.jsi.Function) {
diff --git a/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptObject.swift b/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptObject.swift
index 57bb66234a0dc80b51eedf95b08228781b5eec9b..d58393c76896293f61658db772be180a0fa82669 100644
--- a/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptObject.swift
+++ b/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptObject.swift
@@ -8,7 +8,7 @@ internal import ExpoModulesJSI_Cxx
  supporting property access, modification, enumeration, prototype manipulation, and function calling.
  */
 public struct JavaScriptObject: JavaScriptType, Sendable, ~Copyable {
-  internal weak let runtime: JavaScriptRuntime?
+  nonisolated(unsafe) internal weak var runtime: JavaScriptRuntime?
   internal var pointee: facebook.jsi.Object
 
   /**
diff --git a/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptPromise.swift b/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptPromise.swift
index 94f10ed7c706bc7144adbf282965b42b9e40a308..d1c1618c6c3e99fb270b88487263846f3f0904f7 100644
--- a/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptPromise.swift
+++ b/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptPromise.swift
@@ -12,7 +12,7 @@ internal import ExpoModulesJSI_Cxx
 public struct JavaScriptPromise: JavaScriptType, ~Copyable {
   private typealias PromiseContinuation = CheckedContinuation<JavaScriptValue.Ref, any Error>
 
-  private weak let runtime: JavaScriptRuntime?
+  nonisolated(unsafe) private weak var runtime: JavaScriptRuntime?
   private let object: JavaScriptObject
   private let deferredPromise = DeferredPromise()
 
diff --git a/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptTypedArray.swift b/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptTypedArray.swift
index c8caee56f1dd335692f5edaae0ad3c703b8129ad..b1747f62d0118bf4314f60bdba13916e47ee604f 100644
--- a/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptTypedArray.swift
+++ b/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptTypedArray.swift
@@ -4,7 +4,7 @@ internal import jsi
 internal import ExpoModulesJSI_Cxx
 
 public struct JavaScriptTypedArray: ~Copyable {
-  internal weak let runtime: JavaScriptRuntime?
+  nonisolated(unsafe) internal weak var runtime: JavaScriptRuntime?
   internal let pointee: facebook.jsi.Object
 
   /**
diff --git a/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptValue.swift b/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptValue.swift
index ba58f9e1a59038ed9f6ee4a3bf4a48569ad657dc..efcc9700699aa9f656d8041936b433e800155439 100644
--- a/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptValue.swift
+++ b/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptValue.swift
@@ -9,7 +9,7 @@ internal import ExpoModulesJSI_Cxx
  and stored in containers that don't support non-copyable types etc.
  */
 public final class JavaScriptValue: JavaScriptType, Equatable, Escapable, Error {
-  internal weak let runtime: JavaScriptRuntime?
+  nonisolated(unsafe) internal weak var runtime: JavaScriptRuntime?
   internal let pointee: facebook.jsi.Value
 
   /**
diff --git a/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptWeakObject.swift b/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptWeakObject.swift
index 2907dad8f988ddb0baf22db9a18bc19b74c3f9ff..c9f91812bb608d69d3df9350f70765132f221710 100644
--- a/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptWeakObject.swift
+++ b/apple/Sources/ExpoModulesJSI/Runtime/Values/JavaScriptWeakObject.swift
@@ -7,7 +7,7 @@ internal import jsi
  the object is eligible for GC. Method names are inspired by C++ `std::weak_ptr`.
  */
 public struct JavaScriptWeakObject: JavaScriptType, ~Copyable {
-  internal weak let runtime: JavaScriptRuntime?
+  nonisolated(unsafe) internal weak var runtime: JavaScriptRuntime?
   internal let pointee: facebook.jsi.WeakObject
 
   /**
